### PR TITLE
Fix semantic segmentation mask offset from sub-pixel LetterBox padding

### LINKED ISSUE #24617: Mask misalignment occurs during semantic segmentation inference/prediction.
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Predict

### Bug

When the input image resolution is 1920×1080, the preprocessing pipeline resizes the image to 640×360 and pads 12 pixels on both the top and bottom, yielding an actual model input size of 640×384. However, the semantic segmentation model produces an output of shape nc × 80 × 48. During post-processing, the padding is first computed; yet due to the 8× downsampling, the padding becomes 1.5 pixels on each side in the output feature space. After rounding, this results in 1 pixel of padding on top and 2 pixels on the bottom, ultimately causing a 0.5-pixel downward shift of the mask. When mapped back to the original image resolution, this displacement becomes 0.5 × 8 × 3 = 12 pixels, leading to a significant and noticeable mask offset.

### Environment

Ultralytics 8.4.55  Python-3.12.12 torch-2.10.0+cpu CPU (AMD RYZEN AI MAX+ 395 w/ Radeon 8060S)
Setup complete  (32 CPUs, 27.6 GB RAM, 527.5/924.5 GB disk)

OS                     Windows-11-10.0.26200-SP0
Environment            Windowstralytics.com/solutions/
Python                 3.12.12ty.ultralytics.com
Install                git.com/ultralytics/ultralytics
Path                   C:\Users\lijiq\PycharmProjects\ultralytics\ultralytics
RAM                    27.65 GB
Disk                   527.5/924.5 GBrojects\ultralytics>
CPU                    AMD RYZEN AI MAX+ 395 w/ Radeon 8060S
CPU count              32
GPU                    None
GPU count              None
CUDA                   None

numpy                   2.4.2>=1.23.0
matplotlib              3.10.8>=3.3.0
opencv-python           4.13.0.92>=4.6.0
pillow                  12.1.0>=7.1.2
pyyaml                  6.0.3>=5.3.1
requests                2.32.5>=2.23.0
scipy                   1.17.0>=1.4.1
torch                   2.10.0>=1.8.0
torch                   2.10.0!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision             0.25.0>=0.9.0
psutil                  7.2.2>=5.8.0
polars                  1.38.1>=0.20.0
ultralytics-thop        2.0.18>=2.0.18

### Minimal Reproducible Example

You can resize any image from Cityscapes to 1920×1080, feed it into the pretrained model for inference, and the same result should be observed.

### Additional

_No response_

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by lmycross**
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
🛠️ This PR improves semantic segmentation postprocessing by resizing predictions more carefully before removing padding, leading to cleaner and more accurate masks on original images.

### 📊 Key Changes
- Added `torch.nn.functional as F` to use interpolation during semantic prediction postprocessing.
- Updated `semantic/predict.py` so segmentation outputs are first converted to float and expanded with `unsqueeze(0)`.
- Added a step to upsample predictions to the model input resolution with `F.interpolate(..., mode="bilinear")` when needed.
- Kept `ops.scale_masks(...)` for final resizing back to the original image, but now after aligning predictions to the input size first.
- Included comments explaining that this avoids padding-related sub-pixel asymmetry during mask cropping.

### 🎯 Purpose & Impact
- 🎯 Fixes a subtle mask alignment issue in semantic segmentation predictions, especially around image borders and padded regions.
- 🖼️ Produces more accurate semantic masks when mapping model outputs back to the original image size.
- 📏 Makes postprocessing more consistent when prediction resolution does not exactly match the input image resolution.
- 👥 Benefits users by improving output quality without changing the public API or requiring workflow changes.
- 🚀 Helps make semantic segmentation results in Ultralytics more reliable for real-world inference and visualization.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @lmycross, thank you for submitting this `ultralytics/ultralytics` 🚀 PR! This is an automated message to help with the review process—an engineer will assist you shortly. For this semantic segmentation fix 🩹, please review the checklist below to help ensure a smooth integration:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — APPROVED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The change looks clean and well-scoped. I did not find any clear bugs or regressions in the diff based on the visible context.

**Comment by UltralyticsAssistant** [comment]
Merged with thanks, @lmycross! 🎉

“Details make perfection, and perfection is not a detail.” — Leonardo da Vinci

This update is a great example of that idea in practice: carefully aligning semantic predictions before padding removal improves mask accuracy where it matters most, especially around borders and original-image remapping. Thanks for helping make Ultralytics semantic segmentation more reliable and polished for everyone.
